examples/connect-latency.bpf.c

```diff
@@ -0,0 +1,133 @@
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include <bpf/bpf_core_read.h>
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+#define MAX_LATENCY_SLOT 26
+
+struct connect_start_key_t {
+    u64 pid_tgid;
+};
+
+struct connect_start_val_t {
+    u64 ts;
+    int addrlen;
+    u32 d_ip;    // Destination IPv4 address
+    u16 d_port;  // Destination port number
+};
+
+struct connect_latency_key_t {
+    u32 d_ip;    // Destination IPv4 address
+    u16 d_port;  // Destination port number
+    u64 slot;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, 10240);
+    __type(key, struct connect_start_key_t);
+    __type(value, struct connect_start_val_t);
+} connect_start SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(max_entries, MAX_LATENCY_SLOT + 2);
+    __type(key, struct connect_latency_key_t);
+    __type(value, u64);
+} connect_latency_seconds SEC(".maps");
+
+static inline __u16 ntohs(__u16 value) {
+    return ((value & 0x00FF) << 8) | ((value & 0xFF00) >> 8);
+}
+
+SEC("kprobe/__sys_connect")
+int BPF_KPROBE(kprobe__sys_connect, int sockfd, const struct sockaddr *addr, int addrlen)
+{
+    struct sockaddr sa;
+    struct connect_start_val_t start_val = {};
+
+    bpf_probe_read(&sa, sizeof(sa), addr);
+
+    if (sa.sa_family == 1) {
+        return 0;  // Ignore UNIX domain sockets
+    }
+
+    if (addrlen == sizeof(struct sockaddr_in)) {
+        struct sockaddr_in v4;
+        bpf_probe_read(&v4, sizeof(v4), addr);
+        start_val.d_ip = v4.sin_addr.s_addr;
+        start_val.d_port = v4.sin_port;
+    } else if (addrlen == sizeof(struct sockaddr_in6)) {
+	const char debug_str[] = "This is ipv6!\n";
+	bpf_trace_printk(debug_str, sizeof(debug_str));
+        struct sockaddr_in6 v6;
+        bpf_probe_read(&v6, sizeof(v6), addr);
+
+        if (BPF_CORE_READ(&v6.sin6_addr.in6_u, u6_addr32[0]) == 0x00000000 &&
+            BPF_CORE_READ(&v6.sin6_addr.in6_u, u6_addr32[1]) == 0x00000000 &&
+            BPF_CORE_READ(&v6.sin6_addr.in6_u, u6_addr32[2]) == 0x0000FFFF) {
+
+            start_val.d_ip = BPF_CORE_READ(&v6.sin6_addr.in6_u, u6_addr32[3]);
+            start_val.d_port = v6.sin6_port;
+        } else {
+            const char debug_str[] = "This is native ipv6, I'm giving up!\n";
+            bpf_trace_printk(debug_str, sizeof(debug_str));
+            return 0;
+        }
+    } else {
+        const char debug_str[] = "Unexpected addrlen: %d, address family: %d\n";
```

Comment:
We generally don't leave debug statements around.

Reply:
I would recommend `bpf_printk()` when you debug locally:

* https://nakryiko.com/posts/bpf-tips-printk/